lib/client.js

```diff
@@ -80,56 +81,70 @@ class Metacom extends EventEmitter {
     };
   }
 
-  async message(data) {
-    if (data === '{}') return;
+  async message(msg) {
+    if (msg === '{}') return;
     this.lastActivity = Date.now();
-    let packet;
-    try {
-      packet = JSON.parse(data);
-    } catch {
-      return;
-    }
-    const { type, id, method } = packet;
-    if (id) {
-      if (type === 'callback') {
-        const promised = this.calls.get(id);
-        if (!promised) return;
-        const [resolve, reject, timeout] = promised;
-        this.calls.delete(id);
-        clearTimeout(timeout);
-        if (packet.error) {
-          reject(new MetacomError(packet.error));
-          return;
-        }
-        resolve(packet.result);
-      } else if (type === 'event') {
-        const [unit, name] = method.split('/');
-        const metacomUnit = this.api[unit];
-        if (metacomUnit) metacomUnit.emit(name, packet.data);
-      } else if (type === 'stream') {
-        const { name, size, status } = packet;
-        const stream = this.streams.get(id);
-        if (name && typeof name === 'string' && Number.isSafeInteger(size)) {
-          if (stream) {
-            console.error(new Error(`Stream ${name} is already initialized`));
-          } else {
-            const stream = new MetaReadable({ id, name, size });
-            this.streams.set(id, stream);
-          }
-        } else if (!stream) {
-          console.error(new Error(`Stream ${id} is not initialized`));
-        } else if (status === 'end') {
-          await stream.close();
-          this.streams.delete(id);
-        } else if (status === 'terminate') {
-          await stream.terminate();
-          this.streams.delete(id);
-        } else {
-          console.error(new Error('Stream packet structure error'));
-        }
+    const packet = metautil.jsonParse(msg);
+    const parsed = protocol.deserialize(packet);
+    if (!parsed) throw new Error(''); // TODO - how this should be handled?
+    const { type, data } = parsed;
+    const handlers = {
+      error: this.#errorHandler,
+      callback: this.#callbackHandler,
+      event: this.#eventHandler,
+      stream: this.#streamHandler,
+    };
+    const handler = handlers[type];
+    if (!handler) throw new Error(''); // TODO - how this should be handled?
+    await handler(data);
```

Comment:
To make this work all handlers should be public methods, since `[]` doesn't work for private methods (e.g. ```this[`#${type}Handler`]``` won't work). Having those as private makes future changes easier since you have guaranties that only `Client` can access it.

Reply:
It's better to extract all this handlers outside class and pass this to their first parameter